validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java

```diff
@@ -279,135 +277,83 @@ private void produceAttestations(
-   *
-   * @param state
-   * @param block
-   * @param proposer
-   * @return
-   * @see
-   *     <a>https://github.com/ethereum/eth2.0-specs/blob/v0.8.1/specs/validator/0_beacon-chain-validator.md#signature</a>
-   */
-  private BLSSignature getBlockSignature(
-      BeaconState state, BeaconBlock block, BLSPublicKey proposer) {
-    Bytes domain =
-        get_domain(
-            state,
-            Constants.DOMAIN_BEACON_PROPOSER,
-            BeaconStateUtil.compute_epoch_at_slot(block.getSlot()));
+  private void createBlockIfNecessary(
+      BeaconStateWithCache previousState, BeaconBlock previousBlock) {
+    // TODO - we shouldn't assume that we're just incrementing the slot by 1
+    final UnsignedLong newSlot = previousState.getSlot().plus(UnsignedLong.ONE);
 
-    Bytes32 blockRoot = block.signing_root("signature");
+    // Check if we should be proposing
+    final BLSPublicKey proposer = blockCreator.getProposerForSlot(previousState, newSlot);
+    if (!validators.containsKey(proposer)) {
+      // We're not proposing now
+      return;
+    }
 
-    return getSignature(blockRoot, domain, proposer);
+    BeaconBlock newBlock;
+    try {
+      // Collect attestations to include
+      SSZList<Attestation> attestations = getAttestationsForSlot(newSlot);
+      // Collect slashing to include
+      final SSZList<ProposerSlashing> slashingsInBlock = getSlashingsForBlock(previousState);
+      // Collect deposits
+      final SSZList<Deposit> deposits = getDepositsForBlock();
+
+      final Signer signer = getSigner(proposer);
+      final Bytes32 parentRoot = previousBlock.signing_root("signature");
+      newBlock =
+          blockCreator.createNewBlock(
+              signer, newSlot, previousState, parentRoot, attestations, slashingsInBlock, deposits);
+
+      this.eventBus.post(newBlock);
+      STDOUT.log(Level.DEBUG, "Local validator produced a new block");
+
+      if (validators.get(proposer).isNaughty()) {
+        final BeaconBlock naughtyBlock =
+            blockCreator.createEmptyBlock(signer, newSlot, previousState, parentRoot);
+        this.eventBus.post(naughtyBlock);
+      }
+    } catch (StateTransitionException e) {
+      STDOUT.log(Level.WARN, "Error during block creation " + e.toString());
+    }
   }
 
-  private BeaconBlock createInitialBlock(BeaconStateWithCache state, BeaconBlock oldBlock) {
-    Bytes32 blockRoot = oldBlock.signing_root("signature");
-    SSZList<Attestation> current_attestations = new SSZList<>(Attestation.class, MAX_ATTESTATIONS);
-    final Bytes32 MockStateRoot = Bytes32.ZERO;
-
-    if (state
-            .getSlot()
-            .compareTo(
-                UnsignedLong.valueOf(
-                    Constants.GENESIS_SLOT + Constants.MIN_ATTESTATION_INCLUSION_DELAY))
+  private SSZList<Attestation> getAttestationsForSlot(final UnsignedLong slot) {
+    SSZList<Attestation> attestations = BeaconBlockBodyLists.createAttestations();
```

Comment:
This might be unnecessary here since blockAttestationsPool.getAggreggatedAttestationsForBlockAtSlot returns an sszList.